### PR TITLE
Gravity Gun

### PR DESCRIPTION
**Pull request by silicons**
Gravity guns:

12 second firing delay. Var-editable by admins. This delay might need changes
Doesn't need to recharge
Printable from R&D for a high research and material cost
Two modes: Attract, and repulse
During flight the projectile will apply the effect to a weakened extent every 3 tiles
On hit, will apply the effect to a 5-6 tile radius near-instantly. This is var-editable by admins. 

Power varedits will be limited to around 15 so server crashing doesn't happen from an admin "having fun" and setting the power to 300 which would throw every item on the zlevel and apply a laggy overlay to everything

Sprites are very bad so if anyone wants to give me better sprites to go on them that'd be nice

Balance concerns should be commented below :)

:cl:
rscadd: Gravity Guns! Two mode guns that fire either attraction or repulsion bolts. Available from R&D for a high price!
/:cl:

### CODE REVIEW

**Comment by kevinz000** [comment]
inbefore yelled at for adding expensive, somewhat balanced fun to R&D.. and forgot changelog fuck hold on

**Comment by Cheridan** [comment]
isn't this just TK but easier to get

**Comment by kevinz000** [comment]
Uuh very, very, very expensive TK, as well as not being able to attack anything as it uses step_towards/away instead of throw_at
in non code talk it means you can't use it to launch 30 spears at people with it. long delay too, can always up delay/lower AOE.

**Comment by Core0verload** [comment]
Seems like useless.

**Comment by kevinz000** [comment]
>useless>
Fun

**Comment by Ausops** [comment]
someone quickly check to see if you can use this gun to self-trigger the rd's armor :^)

**Comment by kevinz000** [comment]
No. Yes, i did try.

File: code/modules/projectiles/guns/energy/special.dm [hunk 1]
```diff
@@ -207,6 +207,14 @@
 	origin_tech = "combat=3;materials=4;magnets=3;plasmatech=3;engineering=2"
 	ammo_type = list(/obj/item/ammo_casing/energy/plasma/adv)
 
+/obj/item/weapon/gun/energy/gravity_gun
+	name = "gravitron beamgun"
```
Comment (xxalpha) [inline code/modules/projectiles/guns/energy/special.dm]: zero point energy field manipulator

Comment (kevinz000) [inline code/modules/projectiles/guns/energy/special.dm]: isn't there a rule or something about copyrights? And that would be from HL2

Comment (xxalpha) [inline code/modules/projectiles/guns/energy/special.dm]: Valve didn't invent gravity.

**Comment by Fox-McCloud** [comment]
Not saying this is balanced or not, but this is what science needs; more whacky/unique/weird weapons and less "does the same thing as another weapon but self recharges"

File: code/modules/research/designs/weapon_designs.dm [hunk 2]
```diff
@@ -204,6 +204,16 @@
 	build_path = /obj/item/weapon/gun/energy/wormhole_projector
 	category = list("Weapons")
 
+/datum/design/gravitygun
+	name = "Gravitron Blaster"
+	desc = "A dual-mode projector that blasts high powered gravitrons."
+	id = "gravitygun"
+	req_tech = list("combat" = 6, "materials" = 7, "bluespace" = 5, "powerstorage" = 5)
```
Comment (PKPenguin321) [inline code/modules/research/designs/weapon_designs.dm]: >combat 6 for weaker telekinesis

Comment (kevinz000) [inline code/modules/research/designs/weapon_designs.dm]: >gimmick gun, that's better for combat then you think

**Comment by kevinz000** [comment]
>making it throw things make it op
Making it just move the objects makes it useless

**Comment by Shadowlight213** [comment]
@kevinz000 need to add origin tech to the gun

**Comment by kevinz000** [comment]
Oh woops, sure

**Comment by kevinz000** [comment]
Any suggestions to make it a weapon but still be balanced? Throwing things is pretty much out of the question due to the probability of instakilling people with masses of flying toolboxes...

**Comment by PKPenguin321** [comment]
can you throw people with it?

**Comment by Alek2ander** [comment]
:-1:
Cannot recreate Ravenholm with this kid-friendly replica.

File: code/modules/projectiles/guns/energy/special.dm [hunk 3]
```diff
@@ -207,6 +207,15 @@
 	origin_tech = "combat=3;materials=4;magnets=3;plasmatech=3;engineering=2"
 	ammo_type = list(/obj/item/ammo_casing/energy/plasma/adv)
 
+/obj/item/weapon/gun/energy/gravity_gun
+	name = "Zero-Point Energy Manipulator"
```
Comment (xxalpha) [inline code/modules/projectiles/guns/energy/special.dm]: Energy _Field_

**Comment by kevinz000** [comment]
I mean this could be used to pull people into things, like flashbangs, or literally scatter a lynch mob with a gravitational bowling ball..?

**Comment by Shadowlight213** [comment]
Op admin version for literally recreating ravenholm pls.

**Comment by Iamgoofball** [comment]
yeah can you please make it use throw_at()

**Comment by kevinz000** [comment]
It now throws things if it's in close proximity to the projectile on hit. Making that function able to be varedited will come soon :P 
Seems to be a bit laggy though...

File: code/modules/projectiles/projectile/special.dm [hunk 4]
```diff
@@ -168,6 +168,89 @@
...
+			if(A.anchored)
+				continue
+			step_towards(A,T)
+		rangecooldown = 3
+	rangecooldown--
+
+/obj/item/projectile/gravipulse/alt/on_hit()
+	T = get_turf(src)
+	for(var/atom/movable/A in orange(T,power))
+		if(A == src)
```
Comment (KorPhaeron) [inline code/modules/projectiles/projectile/special.dm]: This exempts the gun, but not the person firing the gun. Did you intend this?

File: code/modules/projectiles/projectile/special.dm [hunk 5]
```diff
@@ -168,6 +168,89 @@
...
+			continue
+		if(A.anchored)
+			continue
+		if(A in range(T,throwpower))
+			A.throw_at(T,throwpower+1,1)
+			continue
+		for(var/iter=0 to power)
+			step_towards(A,T)
+			sleep(1)
```
Comment (KorPhaeron) [inline code/modules/projectiles/projectile/special.dm]: The fuck is this nonsense no wonder it lags

**Comment by KorPhaeron** [comment]
Okay I finally looked at this and you did it in a very odd way, please ping me on IRC I'll help you rewrite it Lag Free(TM)

File: code/modules/projectiles/projectile/special.dm (hunk 4 above)
Comment (kevinz000) [inline code/modules/projectiles/projectile/special.dm]: This is supposed to make the projectile not affect ITSELF.

Comment (KorPhaeron) [inline code/modules/projectiles/projectile/special.dm]: Oh whoops yeah, projectile. Still, exempting the shooter might be a good idea

**Comment by KorPhaeron** [comment]
http://piratepad.net/z9wqHJeIBA

Here you go mekhi

**Comment by kevinz000** [comment]
Ayy thanks now for me to get to a computer

**Comment by kevinz000** [comment]
@KorPhaeron  Your code makes things just spin in place.

**Comment by KorPhaeron** [comment]
For throwing towards or away?

**Comment by kevinz000** [comment]
Fixed. now uh, what the hell is this conflict

**Comment by KorPhaeron** [comment]
New energy gun sprites, it's a dmi conflict

**Comment by kevinz000** [comment]
I feel like I fucked up somewhere. But I don't know where.

**Comment by kevinz000** [comment]
ok, I almost overwrote the mini egun sprites but I fixed it :)

**Comment by kevinz000** [comment]
@KorPhaeron seems to be working fine, anything else?

**Comment by KorPhaeron** [comment]
1k line changes in arcade.dm?

**Comment by kevinz000** [comment]
@KorPhaeron  No idea how that happened but I'm pretty sure it's unchanged.

File: code/modules/projectiles/projectile/special.dm [hunk 6]
```diff
@@ -168,6 +168,53 @@
...
+	damage = 0
+	damage_type = BRUTE
+	nodamage = 1
+	color = "#33CCFF"
+	var/turf/T
+	var/power = 4
+
+/obj/item/projectile/gravipulse/New(var/obj/item/ammo_casing/energy/gravipulse/C)
+	if(C) //Hard-coded maximum power so servers can't be crashed by trying to throw the entire Z level's items
+		power = min(C.gun.power, 15)
```
Comment (phil235) [inline code/modules/projectiles/projectile/special.dm]: ..()

File: code/modules/projectiles/projectile/special.dm [hunk 7]
```diff
@@ -168,6 +168,53 @@
...
+
+/obj/item/ammo_casing/energy/gravipulse/New(var/obj/item/weapon/gun/energy/gravity_gun/G)
+	gun = G
+
+/obj/item/projectile/gravipulse/on_hit()
```
Comment (phil235) [inline code/modules/projectiles/projectile/special.dm]: . = ..()

File: code/modules/projectiles/projectile/special.dm [hunk 8]
```diff
@@ -168,6 +168,53 @@
...
+	for(var/turf/F in range(T,power))
+		var/obj/effect/overlay/gravfield = new /obj/effect/overlay{icon='icons/effects/effects.dmi'; icon_state="shieldsparkles"; mouse_opacity=0; density=0}()
+		F.overlays += gravfield
+		spawn(5)
+		F.overlays -= gravfield
+
+/obj/item/projectile/gravipulse/alt
+	color = "#FF6600"
+
+/obj/item/projectile/gravipulse/alt/on_hit()
```
Comment (phil235) [inline code/modules/projectiles/projectile/special.dm]: . = ..()

**Comment by kevinz000** [comment]
@KorPhaeron  should be fixed.

**Comment by KorPhaeron** [comment]
Would conflict with pixel projectiles

**Comment by kevinz000** [comment]
Uhhhhhhhhhhhhhhhhhhhh what ok whatever will come on irc later

**Comment by PKPenguin321** [comment]
the solution is to not add pixel projectiles tbh

**Comment by optimumtact** [comment]
This looks dead and has conflicts - make sure to also take note of the items that phil mentioned if you work on this again.

**Pull request closed without merging**